### PR TITLE
[SPARK-38360][SQL][AVRO][SS][FOLLOWUP] Replace `TreeNode.collectFirst` + `isDefined/isEmpty` with `exists`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This is a follow up of SPARK-38360 to simplify code patterns related to `TreeNode.collectFirst`, the simplified rules are as follows:

- `treeNode.collectFirst(condition).isDefined` -> `treeNode.exists(condition)`
- `treeNode.collectFirst(condition).isEmpty` -> `!treeNode.exists(condition)`

### Why are the changes needed?
Code simplification

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA

### CODE REVIEW

**Comment by LuciferYang** [comment]
@zhengruifeng 

this is the follow of SPARK-38360 refer to https://github.com/apache/spark/pull/35694#pullrequestreview-905784925

also cc @cloud-fan @viirya @HyukjinKwon

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaContinuousSourceSuite.scala [hunk 1]
```diff
@@ -212,13 +212,14 @@ class KafkaContinuousSourceTopicDeletionSuite extends KafkaContinuousTest {
...
         eventually(timeout(streamingTimeout)) {
           assert(
-            query.lastExecution.executedPlan.collectFirst {
-              case scan: ContinuousScanExec
-                if scan.stream.isInstanceOf[KafkaContinuousStream] =>
-                scan.stream.asInstanceOf[KafkaContinuousStream]
-            }.exists { stream =>
-              // Ensure the new topic is present and the old topic is gone.
-              stream.knownPartitions.exists(_.topic == topic2)
+            query.lastExecution.executedPlan.exists {
```
Comment (zhengruifeng) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaContinuousSourceSuite.scala:215]: I guess this logic is changed?

suppose there are two `ContinuousScanExec` and the first one do not satisify `stream.knownPartitions.exists(_.topic == topic2)` but the second one satisify it.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaContinuousTest.scala [hunk 2]
```diff
@@ -45,11 +45,15 @@ trait KafkaContinuousTest extends KafkaSourceTest {
     testUtils.addPartitions(topic, newCount)
     eventually(timeout(streamingTimeout)) {
       assert(
-        query.lastExecution.executedPlan.collectFirst {
-          case scan: ContinuousScanExec
-              if scan.stream.isInstanceOf[KafkaContinuousStream] =>
-            scan.stream.asInstanceOf[KafkaContinuousStream]
-        }.exists(_.knownPartitions.size == newCount),
+        query.lastExecution.executedPlan.exists {
```
Comment (zhengruifeng) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaContinuousTest.scala:48]: ditto

**Comment by zhengruifeng** [comment]
hmm, there are some `case _ => false` added since collectFirst use a partialfunction.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaContinuousSourceSuite.scala (hunk 1 above)
Comment (LuciferYang) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaContinuousSourceSuite.scala]: You are right. The logic has changed, [0031b96](https://github.com/apache/spark/pull/35833/commits/0031b96ae460fa0adf42ae14d0ea6c66b5df0501) revert this changed.

Thanks ~

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaContinuousTest.scala (hunk 2 above)
Comment (LuciferYang) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaContinuousTest.scala]: [0031b96](https://github.com/apache/spark/pull/35833/commits/0031b96ae460fa0adf42ae14d0ea6c66b5df0501) revert this changed.

**Comment by LuciferYang** [comment]
> hmm, there are some `case _ => false` added since collectFirst use a partialfunction.

Yes, the code has not been reduced :(

**Review by cloud-fan — APPROVED** [review]

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
Thanks, merged to master!

**Comment by LuciferYang** [comment]
thanks all

**Pull request closed without merging**